presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -227,62 +228,47 @@ private static boolean isCovariantParameterPosition(String firstTypeBase, int po
-            return true;
-        }
-
-        if (actualType.getBase().equals(StandardTypes.DECIMAL) && expectedType.getBase().equals(StandardTypes.DECIMAL)) {
-            long actualPrecision = actualType.getParameters().get(0).getLongLiteral();
-            long expectedPrecision = (long) expectedType.getParameters().get(0).getLongLiteral();
-            long actualScale = (long) actualType.getParameters().get(1).getLongLiteral();
-            long expectedScale = (long) expectedType.getParameters().get(1).getLongLiteral();
 
-            if (actualPrecision <= Decimals.MAX_SHORT_PRECISION ^ expectedPrecision <= Decimals.MAX_SHORT_PRECISION) {
-                return false;
-            }
+        if (!canCoerce(source, result)) {
+            return false;
+        }
 
-            return actualScale == expectedScale && actualPrecision <= expectedPrecision;
+        if (source instanceof VarcharType && result instanceof VarcharType) {
+            return true;
         }
 
-        if (actualType.getBase().equals(expectedType.getBase()) &&
-                actualType.getParameters().size() == expectedType.getParameters().size()) {
-            for (int i = 0; i < actualType.getParameters().size(); i++) {
-                if (!isCovariantParameterPosition(actualType.getBase(), i)) {
-                    return false;
-                }
+        if (source instanceof DecimalType && result instanceof DecimalType) {
+            DecimalType sourceDecimal = (DecimalType) source;
+            DecimalType resultDecimal = (DecimalType) result;
+            boolean sameDecimalSubtype = (sourceDecimal.isShort() && resultDecimal.isShort())
+                    || (!sourceDecimal.isShort() && !resultDecimal.isShort());
+            boolean sameScale = sourceDecimal.getScale() == resultDecimal.getScale();
+            boolean sourcePrecisionIsLessOrEqualToResultPrecision = sourceDecimal.getPrecision() <= resultDecimal.getPrecision();
+            return sameDecimalSubtype && sameScale && sourcePrecisionIsLessOrEqualToResultPrecision;
+        }
 
-                TypeSignatureParameter actualParameter = actualType.getParameters().get(i);
-                TypeSignatureParameter expectedParameter = expectedType.getParameters().get(i);
-                if (actualParameter.equals(expectedParameter)) {
-                    continue;
-                }
+        String sourceTypeName = source.getTypeSignature().getBase();
+        String resultTypeName = result.getTypeSignature().getBase();
 
-                Optional<TypeSignature> actualParameterSignature = actualParameter.getTypeSignatureOrNamedTypeSignature();
-                Optional<TypeSignature> expectedParameterSignature = expectedParameter.getTypeSignatureOrNamedTypeSignature();
-                if (!actualParameterSignature.isPresent() || !expectedParameterSignature.isPresent()) {
+        if (sourceTypeName.equals(resultTypeName)) {
```

Comment:
this is true for ARRAY, MAP and ROW. But it is not general truth. We can leave it for now. As we do not have more types taking types as parameters.